### PR TITLE
[BEAM-4242] Fixes O(n) complexity of Wait.on() due to combiner lifting

### PR DESCRIPTION
**Pull request by jkff**
* Fixes Wait implementation to not depend on combiner lifting happening
* Fixes the condition for combiner lifting to only account for the trigger in streaming

R: @chamikaramj 
CC: @mairbek

### CODE REVIEW

**Comment by jkff** [comment]
Run Java PreCommit

**Comment by jkff** [comment]
Precommit failure was an existing failing test; rebased anyway to pick up the blacklisting of that test, PTAL.

**Comment by jkff** [comment]
Run Java PreCommit

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java [hunk 1]
```diff
@@ -844,13 +844,15 @@ public void translate(GroupByKey transform, TranslationContext context) {
...
-                    || !windowingStrategy.getWindowFn().assignsToOneWindow()
-                    || (isStreaming && !transform.fewKeys())
-                    // TODO: Allow combiner lifting on the non-default trigger, as appropriate.
-                    || !(windowingStrategy.getTrigger() instanceof DefaultTrigger);
-            stepContext.addInput(PropertyNames.DISALLOW_COMBINER_LIFTING, disallowCombinerLifting);
+            boolean allowCombinerLifting =
+                windowingStrategy.getWindowFn().isNonMerging()
+                    && windowingStrategy.getWindowFn().assignsToOneWindow();
+            if (isStreaming) {
+              allowCombinerLifting &= !transform.fewKeys();
```
Comment (chamikaramj) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java]: transform.fewKeys() (without !) ?

Comment (jkff) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java]: Oops, thanks.

**Pull request merged**